src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs

```diff
@@ -140,12 +142,108 @@ public ResultWrapper<Address> eth_coinbase()
+        {
+            while (headBlockNumber >= genesisBlockNumber) //should this be latestBlock or headBlock?
+            {
+                Transaction[] transactions = _blockFinder.FindBlock(headBlockNumber)!.Transactions;
+                if (transactions.Length > 0)
+                {
+                    latestGasPrice = transactions[^1].GasPrice;
+                    return;
+                }
+
+                headBlockNumber--;
+            }
+
+            latestGasPrice = 1; //do we want to throw an error if there are no transactions? Do we want to set lastPrice to 1 when we cannot find latestPrice in tx?
         }
+        
+        [Todo("Gas pricer to be implemented")]
+        public ResultWrapper<UInt256?> eth_gasPrice(UInt256? ignoreUnder = null)
+        {
+            Block headBlock = _blockFinder.FindHeadBlock();
+            Block genesisBlock = _blockFinder.FindGenesisBlock();
+            if (headBlock == null || genesisBlock == null)
+            {
+                return ResultWrapper<UInt256?>.Fail("The head block or genesis block had a null value.");
+            }
+            Comparer<UInt256> comparer = Comparer<UInt256>.Create(((a, b) =>
+            {
+                int res = a.CompareTo(b);
+                return res == 0 ? 1 : res;
+            })); 
+            SortedSet<UInt256> gasPrices = new(comparer); //allows duplicates
+            long blocksToGoBack = 5;
+            const int percentile = 20;
+            long threshold = blocksToGoBack * 2;
+            UInt256? gasPriceLatest = null;
+            latest_gas_price(headBlock!.Number, genesisBlock!.Number, ref gasPriceLatest);
+            if (gasPriceLatest == null)
+            {
+                return ResultWrapper<UInt256?>.Fail("gasPriceLatest was not set properly.");
+            }
+            long blockNumber = headBlock!.Number; 
 
+            while (blocksToGoBack > 0 && blockNumber > genesisBlock!.Number - 1) //else, go back "blockNumber" valid gas prices from genesisBlock
+            {
+                Block? foundBlock = _blockFinder.FindBlock(blockNumber);
+                if (foundBlock != null)
+                {
+                    bool result = add_transactions_from_block_to_set(foundBlock, ref gasPrices, (UInt256) gasPriceLatest, ignoreUnder);
+                    if (result || gasPrices.Count + blocksToGoBack >= threshold) //if we only added one transaction, we don't reduce blocksToGoBack if second condition holds
+                    {
+                        blocksToGoBack--;
+                    }
+                }
+                blockNumber--;
+            }
+
+            int finalIndex = (int) Math.Round(((gasPrices.Count - 1) * ((float) percentile / 100)));
+            foreach (UInt256 gasPrice in gasPrices.Where(_ => finalIndex-- <= 0))
+            {
+                gasPriceLatest = gasPrice;
+                break;
```

Comment:
The logic is quite hard to follow, can this be simplified by better naming or more submethods?